peer/hashring32/ring.go

```diff
@@ -0,0 +1,162 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package hashring32
+
+import (
+	"strconv"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/peer/abstractlist"
+	"go.uber.org/yarpc/peer/hashring32/internal/hashring32"
+	"go.uber.org/zap"
+)
+
+// newPeerRing creates a new peerRing with an initial capacity
+func newPeerRing(hashFunc hashring32.HashFunc32, offsetHeader, peerOverrideHeader string, logger *zap.Logger, option ...hashring32.Option) *peerRing {
+	return &peerRing{
+		ring:               hashring32.New(hashFunc, option...),
+		subscribers:        make(map[string]*subscriber),
+		offsetHeader:       offsetHeader,
+		peerOverrideHeader: peerOverrideHeader,
+		logger:             logger,
+	}
+}
+
+type subscriber struct {
+	peer peer.StatusPeer
+}
+
+func (s *subscriber) UpdatePendingRequestCount(int) {}
+
+// peerRing provides a safe way to interact (Add/Remove/Get) with a potentially
+// changing list of peer objects
+// peerRing is NOT Thread-safe, make sure to only call peerRing functions with a lock
+type peerRing struct {
+	ring               *hashring32.Hashring32
+	subscribers        map[string]*subscriber
+	offsetHeader       string
+	peerOverrideHeader string
+	logger             *zap.Logger
+}
+
+var _ abstractlist.Implementation = (*peerRing)(nil)
+
+// shardIdentifier is the interface for an identifier that have a shard property
+type shardIdentifier interface {
+	Identifier() string
+	Shard() string
+}
+
+// Add a string to the end of the peerRing, if the ring is empty
+// it initializes the ring marker
+func (pr *peerRing) Add(p peer.StatusPeer, pid peer.Identifier) abstractlist.Subscriber {
+	sub := &subscriber{peer: p}
+	shardID := getShardID(pid)
+	pr.ring.Add(shardID)
+	pr.subscribers[shardID] = sub
+
+	return sub
+}
+
+// Remove the peer from the ring. Use the subscriber to address the node of the
+// ring directly.
+func (pr *peerRing) Remove(p peer.StatusPeer, pid peer.Identifier, s abstractlist.Subscriber) {
+	sub, ok := s.(*subscriber)
+	if !ok {
+		// Don't panic.
+		return
+	}
+	// Peerlist's responsibility to make sure this never happens.
+	if sub.peer.Identifier() != p.Identifier() {
+		return
+	}
+
+	shardID := getShardID(pid)
+	pr.ring.Add(shardID)
+	pr.subscribers[shardID] = sub
+
+	// validate that given peer is already in the subscriber
+	pr.ring.Remove(shardID)
+	// Peerlist's responsibility to make sure this is thread-safe.
+	delete(pr.subscribers, shardID)
+}
+
+func (pr *peerRing) getPeerOverride(req *transport.Request) peer.StatusPeer {
+	dest, ok := req.Headers.Get(pr.peerOverrideHeader)
+	if !ok {
+		return nil
+	}
+	sub, ok := pr.subscribers[dest]
```

Comment:
are members of the subscribers list always guaranteed to be connected? In the internal version we had 
```
	if !ok || sub.peer.Status().ConnectionStatus != peer.Available {
```

Reply:
Yeah, the new abstract list provides this guarantee. Calling peer.Status() in here would cause a deadlock now.